scio-core/src/main/scala/com/spotify/scio/coders/BeamCoders.scala

```diff
@@ -33,38 +34,49 @@ private[scio] object BeamCoders {
       case _                        => coder
     }
 
+  @inline
+  private def coderElement[T](productCoder: RecordCoder[_])(n: Int): beam.Coder[T] =
+    productCoder.cs(n)._2.asInstanceOf[beam.Coder[T]]
+
+  /** Get coder from an `PCollection[T]`. */
+  def getCoder[T](coll: PCollection[T]): Coder[T] = Coder.beam(unwrap(coll.getCoder))
+
+  /** Get coder from an `SCollection[T]`. */
+  def getCoder[T](coll: SCollection[T]): Coder[T] = getCoder(coll.internal)
+
   /** Get key-value coders from an `SCollection[(K, V)]`. */
   def getTupleCoders[K, V](coll: SCollection[(K, V)]): (Coder[K], Coder[V]) = {
     val coder = coll.internal.getCoder
     val (k, v) = unwrap(coder) match {
-      case c: scio.Tuple2Coder[K, V] => (c.ac, c.bc)
+      case c: scio.Tuple2Coder[K, V] =>
+        (c.ac, c.bc)
       case c: RecordCoder[(K, V)] =>
-        (
-          c.cs.find(_._1 == "_1").get._2.asInstanceOf[beam.Coder[K]],
-          c.cs.find(_._1 == "_2").get._2.asInstanceOf[beam.Coder[V]]
-        )
+        val ac = coderElement[K](c)(0)
+        val bc = coderElement[V](c)(1)
+        (ac, bc)
       case _ =>
         throw new IllegalArgumentException(
           s"Failed to extract key-value coders from Coder[(K, V)]: $coder"
         )
     }
-    (Coder.beam(k), Coder.beam(v))
+    (Coder.beam(unwrap(k)), Coder.beam(unwrap(v)))
```

Comment:
Is calling `unwrap` required everywhere `Coder.beam` is called?

Reply:
No. Most of the usage of `Coder.beam` are not with unwrap like [here](https://github.com/spotify/scio/blob/main/scio-core/src/main/scala/com/spotify/scio/coders/instances/JavaCoders.scala).
We have to unwrap when moving back from materilalized coder back to scio coders. (reuse of internal coders in partial transformation with `keyBy`, `mapValues`, ...)